Allow a stream to skip a lumi if all events processed

wddgit opened the pull request:
#### PR description:

This PR will make the Framework try to direct a stream to skip a lumi if all events have been processed by other streams and the stream has not started the luminosity block yet.

Note that a stream might have started a lumi but not gotten any events yet and then it is too late to skip the lumi. Even after this PR is merged, it will still be possible for a stream to just run begin stream lumi and end stream lumi with no events. 

This new behavior will be a performance improvement when one event takes a very long time to process. Before this PR, the stream processing that event cannot process any subsequent lumis and those lumis cannot complete until after the event completes. If the limit on the number of lumis is reached, all other work might become stuck until the slow event completes.

#### PR validation:

All pre-existing unit Core unit tests pass.

wddgit committed:
Allow a stream to skip a lumi if all events processed

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43522/38104

- This PR adds an extra 48KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/interface/EventProcessor.h  modified in PR(s): #43457
  - File  FWCore/Framework/src/EventProcessor.cc  modified in PR(s): #43457

cmsbuild commented:
A new Pull Request was created by  @wddgit (W. David Dagenhart) for master.

It involves the following packages:

- FWCore/Framework (**core**)


@makortel, @cmsbuild, @Dr15Jones, @smuzaffar can you please review it and eventually sign? Thanks.
@missirol, @makortel this is something you requested to watch as well.
@rappoccio, @antoniovilela, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

wddgit commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ef83f9/36373/summary.html
**COMMIT**: 6c17d4e00cc0b8438a318f18e3b041202e85ff43
**CMSSW**: CMSSW_14_0_X_2023-12-07-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43522/36373/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 2 errors in the following unit tests:

<pre>---> test TestFWCoreFrameworkTransitions had ERRORS
---> test TestFWCoreFrameworkGlobalStreamOne had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [added 74 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-07-1100+ef83f9/60203/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 28 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-07-1100+ef83f9/60203/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3430794
 - DQMHistoTests: [Total failures: 1279](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-07-1100+ef83f9/60203/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3429493
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-07-1100+ef83f9/60203/triggerResults): no differences found

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43522/38165

- This PR adds an extra 52KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/interface/EventProcessor.h  modified in PR(s): #43457
  - File  FWCore/Framework/src/EventProcessor.cc  modified in PR(s): #43457

cmsbuild commented:
Pull request #43522 was updated. @makortel, @Dr15Jones, @smuzaffar, @cmsbuild can you please check and sign again.

wddgit commented:
please test

Just pushed a commit to fix unit test TestFWCoreFrameworkTransitions. Now it passes with this PR. Sometimes that test has a stream that skips a lumi. It does not happen always (seems like very roughly half the time).The test criteria were modified to allow the test to pass if a stream lumi begin or end transition is skipped. It also checks that the expected number of transitions equals the seen number of transitions plus the number skipped, that skipped stream lumi transitions come in matched pairs (a begin and end for the same stream and lumi), and that there is at least one stream begin lumi for every global begin lumi transition.

(There's still one more failing. Working on that next.)

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ef83f9/36448/summary.html
**COMMIT**: 02f05e06c8329f15d3bf306fccd42e699b5df9eb
**CMSSW**: CMSSW_14_0_X_2023-12-12-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36448/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestFWCoreFrameworkGlobalStreamOne had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [removed 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-12-1100+ef83f9/60259/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 12 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-12-1100+ef83f9/60259/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3429858
 - DQMHistoTests: [Total failures: 9](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-12-1100+ef83f9/60259/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3429827
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-12-1100+ef83f9/60259/triggerResults): no differences found

makortel commented on FWCore/Framework/src/EventProcessor.cc:
```diff
@@ -2386,6 +2390,19 @@ namespace edm {
+        continue;
+      }
+      if (iStatus.setStreamStartedLumi(stream)) {
+        // This cannot be the last stream to end the lumi, but we must decrement the counter
+        iStatus.streamFinishedLumi();
```
Would it be worth of asserting the expectation?
```suggestion
        // This cannot be the last stream to end the lumi, but we must decrement the counter
        bool lastStreamToFinish = iStatus.streamFinishedLumi();
        assert(not lastStreamToFinish);
```
Although I'd expect skipping the `iStreamIndex` element, and the `tryToMakeOtherStreamsSkipLumi()` function being called from serial section, to guarantee that.

wddgit replied:
I added the assert and edited the comment also (I'll push it soon with changes related to your other comments).

It cannot be last because the stream associated with ```iStreamIndex``` does not call ```streamEndLumiAsync``` until after ``` tryToMakeOtherStreamsSkipLumi``` returns. Maybe an assert is a good idea. It shouldn't ever fail but if it did and this was the last, then without the assert the Framework would just silently get stuck forever because ```globalEndLumiAsync``` would never get called.

makortel commented on FWCore/Framework/src/LuminosityBlockProcessingStatus.h:
```diff
@@ -70,6 +70,8 @@ namespace edm {
     void setGlobalEndRunHolder(WaitingTaskHolder);
     void globalEndRunHolderDoneWaiting() { globalEndRunHolder_.doneWaiting(std::exception_ptr{}); }
 
+    bool setStreamStartedLumi(unsigned int iStream);
```
I think it would (eventually, say after we've gotten feedback from Chris too on the viability of this approach) be good to add a comment here about the purpose of this function, and the meaning of the return value.

wddgit replied:
I added a comment here.

makortel commented on FWCore/Services/plugins/CheckTransitions.cc:
```diff
@@ -276,7 +310,17 @@ void CheckTransitions::postEndJob() {
   }
 
-  if (orderedSeen.size() != orderedExpected.size()) {
+  if (not(seenGlobalBeginLumi == seenStreamBeginLumi)) {
```
How about
```suggestion
  if (seenGlobalBeginLumi != seenStreamBeginLumi) {
```
? Or did you have reason avoid that?

(same for the if statement few lines below)

wddgit replied:
Should have done it that way in the first place. Thanks. Done.

makortel commented:
To me the approach overall looks viable.

wddgit commented:
I understand the unit test failure in TestFWCoreFrameworkGlobalStreamOne. This test counts transitions including the stream lumi transitions and makes sure the count matches a configured expected number.

This unit test covers many cases, (One,Global,Stream,Limited) x (Analyzer,Producer,Filter) x (several variations of different "Ability" template parameters for each type of module). So it's going to take a while to work through all the cases. The work is similar for all of them. Separate the stream lumi transition count from the count of the other transitions. Appropriately lower the expected values for the other transitions. Separately require the stream lumi transition count to lie between nLumis and nLumis*nStreams (not a single value, but a range). I'll be working my way through all those cases... I've got two working so far.

As far as I can see so far, the Core changes that are still needed are all in the unit test.

makortel commented:
test parameters:
  - enable = threading

makortel commented:
@cmsbuild, please test

Seems like threaded tests could be useful at this point

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ef83f9/36483/summary.html
**COMMIT**: 02f05e06c8329f15d3bf306fccd42e699b5df9eb
**CMSSW**: CMSSW_14_0_X_2023-12-13-1100/el8_amd64_gcc12
**Additional Tests**: THREADING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36483/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestFWCoreFrameworkGlobalStreamOne had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [added 84 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-13-1100+ef83f9/60273/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 20 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-13-1100+ef83f9/60273/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3429858
 - DQMHistoTests: [Total failures: 1205](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-13-1100+ef83f9/60273/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3428631
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-13-1100+ef83f9/60273/triggerResults): no differences found

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43522/38203

- This PR adds an extra 52KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/interface/EventProcessor.h  modified in PR(s): #43457
  - File  FWCore/Framework/src/EventProcessor.cc  modified in PR(s): #43457

cmsbuild commented:
Pull request #43522 was updated. @smuzaffar, @makortel, @Dr15Jones, @cmsbuild can you please check and sign again.

wddgit commented:
please test

I finished all the changes in response to Matti's comments. The third commit at the moment. I'll probably squash it eventually.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ef83f9/36503/summary.html
**COMMIT**: 8ec82a587d05289edc79053cce67b35ddc09f78f
**CMSSW**: CMSSW_14_0_X_2023-12-14-1100/el8_amd64_gcc12
**Additional Tests**: THREADING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36503/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestFWCoreFrameworkGlobalStreamOne had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [removed 9 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-14-1100+ef83f9/60293/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 6 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-14-1100+ef83f9/60293/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3429858
 - DQMHistoTests: [Total failures: 3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-14-1100+ef83f9/60293/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3429833
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-14-1100+ef83f9/60293/triggerResults): no differences found

wddgit commented:
Some comments on this PR (mostly for @Dr15Jones).

First, I would suggest a very careful review of the changes in EventProcessor and LuminosityBlockProcessingStatus (both the header and .cc files). All the rest is unit test updates, maybe not as important. I'm not sure at this point if I'm heading in the right direction or need to be aiming for some different behavior.

This PR as currently written does not affect runs. Probably we could make similar changes for runs, but that is not implemented here yet (I could try to add it if requested or it could be in a later PR or simply not be done).

There are 3 different behaviors under discussion so far:

* Current behavior is every stream sees every lumi.

* This PR will not start a stream on a lumi if all the events are already processed. Note that the same process could be run multiple times and the number of streams that process a particular lumi could vary. Extreme cases are possible. For example, there could be a thousand events and one stream could process them all and only one stream sees the lumi. At the other extreme, there could be zero events in a lumi and still all streams see the lumi. 

* The behavior described in the current documentation is neither of the above:
```
It is possible for the Source to start a new LuminosityBlock and for some of the
Streams to never see that Run [sic] if there are less Events in that LuminosityBlock
then there are Streams in the job. In the case where there are no Events in a
LuminosityBlock, one and only one Stream will see a
beginLuminosityBlock/endLuminosityBlock transition pair.
```

I'm currently still updating the second unit test that was failing ("global" and "limited" cases are done in my working area, I just discovered the "one" cases are OK as is, starting on the "stream" cases now).

Dr15Jones commented on FWCore/Framework/src/LuminosityBlockProcessingStatus.h:
```diff
@@ -103,6 +115,7 @@ namespace edm {
     std::vector<std::shared_ptr<const EventSetupImpl>> eventSetupImpls_;
     WaitingTaskList endIOVWaitingTasks_;
     edm::WaitingTaskHolder globalEndRunHolder_;
+    std::vector<std::atomic<bool>> streamStartedLumi_;
```
I think there might be an easier way. It appears to me that there are now two questions that need to be answered
1. should a stream start processing this lumi
2. once a stream ends processing the lumi, is this the last stream to finish that lumi

I think we could answer these questions by adding only one additional new member
```C++
enum class State {kRunning, kUpdating, kDone};
std::atomic<State> state_{State::kRunning};
```

Then having the following two methods
```C++
bool shouldStreamStartLumi() {
        if (state_ == State::kDone) return false;

   	auto expected = State::kRunning;

	bool changed = false;
	do {
	  changed = state_.compare_exchange_strong(expected, State::kUpdating);
	  if (expected == kDone) return false;
	  expected = kRunning;
	} while(changed == false);

	++nStreamsStillProcessingLumi_;
	state_ = State.kRunning;
	return true;
}
```

```C++
bool streamFinishedLumi() {
	auto expected = State.kRunning;
	bool changed = false;
	do {
	  changed = state_.compare_exchange_strong(expected, State::kUpdating);
	  if (expected == State.kUpdating) expected = State.kRunning;
          //NOTE if expected == State.kDone, that is fine so no need to change to kRunning
	} while(changed ==false);
	auto remaining = --nStreamsStillProcessingLum_;
	state_ = State::kDone;
	return 0 == remaining;	
}
```
state_ is protecting `nStreamsStillProcessingLumi_` so only if we are in kRunning state can the value be incremented and therefore it is safe to decrement and check for 0 as once it reaches 0 it can never increase again. Because of this protection, `nStreamsStillProcessingLumi_` would no longer have to be `std::atomic`.

wddgit replied:
I mostly followed Chris's suggestion. I like that it eliminated the vector of atomics with a single atomic. That is better. The logic is a little more complex with the spin locks and such and probably a few more atomic operations, but memory usage is reduced.

Some modifications on top of Chris's suggestion. I added a new function that mostly does what the suggested version of streamFinishedLumi does and named it noMoreEventsInLumi. It is called once per lumi instead of once per stream per lumi. nStreamsStillProcessingLumi_ remains an atomic, but there is a new member called nStreamsProcessingLumi_ used in the incrementing that is not atomic that gets copied at the correct point. And I changed one of the enum names from kDone to kNoMoreEvents.

The changes are in EventProcessor.cc (the .h file is no longer modified) and LuminosityBlockProcessingStatus.h and .cc. That is the most important part. All the rest is related to unit tests. Existing unit tests should all be working now and I am currently thinking about whether new unit tests are needed and what they ought to do.

Dr15Jones commented on FWCore/Framework/src/EventProcessor.cc:
```diff
@@ -1732,6 +1732,9 @@ namespace edm {
                             streamQueuesInserter_.push(*holder.group(), [this, status, holder, &es]() mutable {
                               for (unsigned int i = 0; i < preallocations_.numberOfStreams(); ++i) {
                                 streamQueues_[i].push(*holder.group(), [this, i, status, holder, &es]() mutable {
+                                  if (!status->setStreamStartedLumi(i)) {
```
This would be replaced with `!status->shouldStreamStartLumi()`

wddgit replied:
Done.

Dr15Jones commented on FWCore/Framework/src/EventProcessor.cc:
```diff
@@ -2256,6 +2259,7 @@ namespace edm {
                 // If appropriate, this will also start the next run.
                 endRunAsync(streamRunStatus_[iStreamIndex], iTask);
               }
+              tryToMakeOtherStreamsSkipLumi(iStreamIndex, *status);
```
this would not be needed (see below)

wddgit replied:
Done.

Dr15Jones commented on FWCore/Framework/src/LuminosityBlockProcessingStatus.cc:
```diff
@@ -14,6 +14,14 @@
 namespace edm {
+
+  LuminosityBlockProcessingStatus::LuminosityBlockProcessingStatus(unsigned int iNStreams)
+      : streamStartedLumi_(iNStreams), nStreamsStillProcessingLumi_(iNStreams) {
```
`nStreamsStillProcessingLumi_` would start at 0

wddgit replied:
Done.

Dr15Jones reviewed (commented):
I propose an alternative implementation.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43522/38301

- This PR adds an extra 92KB to repository

cmsbuild commented:
Pull request #43522 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

wddgit commented on FWCore/Framework/src/LuminosityBlockProcessingStatus.cc:
```diff
@@ -26,6 +34,11 @@ namespace edm {
 
+  bool LuminosityBlockProcessingStatus::setStreamStartedLumi(unsigned int iStream) {
+    bool expected = false;
+    return streamStartedLumi_.at(iStream).compare_exchange_strong(expected, true);
```
This function was deleted entirely.

wddgit commented:
please test

I think this addresses all comments, although there are additional changes on top of what Chris suggested so you should probably take a look again. Existing unit tests should pass now.

wddgit commented:
I'm assuming we're agreed the behavior of this is good since neither of you commented on that. The original implementation, Chris's suggestion, and the version I just pushed should all behave about the same. The changes were all in the implementation details.

I'm also assuming any analogous changes for runs will come in a later PR, if ever.

I guess next week we can discuss how to move forward with this. Unless I missed something, Matti didn't find any problems outside the Framework with the new behavior. I'm trying to think of how to write a new unit test, although that is hard as this behavior of skipping could vary from one process to the next...

cmsbuild commented:
-1

**Failed Tests**: RelVals-THREADING 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ef83f9/36680/summary.html
**COMMIT**: 44994e6e369e4f6aa5fc61c7bd59ac478ab931e9
**CMSSW**: CMSSW_14_0_X_2023-12-28-1100/el8_amd64_gcc12
**Additional Tests**: THREADING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36680/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-THREADING

- [7.3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ef83f9/36680/runTheMatrixTHREADING-results/7.3_CosmicsSPLoose2018)```7.3_CosmicsSPLoose2018/step5_CosmicsSPLoose2018.log```


## Comparison Summary



**Summary**:
 - You potentially [removed 100 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-28-1100+ef83f9/60436/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 20 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-28-1100+ef83f9/60436/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3247277
 - DQMHistoTests: [Total failures: 1216](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-28-1100+ef83f9/60436/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3246039
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-12-28-1100+ef83f9/60436/triggerResults): no differences found

makortel commented:
So 7.3 step 5 (DQM harvesting) crashed with
```
A fatal system signal has occurred: segmentation violation
The following is the call stack containing the origin of the signal.

Thu Dec 28 23:53:50 CET 2023
Thread 2 (Thread 0x149f3acea700 (LWP 1286625) "cmsRun"):
#0  0x0000149f6657a672 in waitpid () from /lib64/libpthread.so.0
#1  0x0000149f61c3d637 in edm::service::cmssw_stacktrace_fork() () from /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36680/CMSSW_14_0_X_2023-12-28-1100/lib/el8_amd64_gcc12/pluginFWCoreServicesPlugins.so
#2  0x0000149f61c4050a in edm::service::InitRootHandlers::stacktraceHelperThread() () from /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36680/CMSSW_14_0_X_2023-12-28-1100/lib/el8_amd64_gcc12/pluginFWCoreServicesPlugins.so
#3  0x0000149f66c04a73 in std::execute_native_thread_routine (__p=0x149f547d34b0) at ../../../../../libstdc++-v3/src/c++11/thread.cc:82
#4  0x0000149f665701ca in start_thread () from /lib64/libpthread.so.0
#5  0x0000149f661dce73 in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x149f656fb640 (LWP 1286602) "cmsRun"):
#0  0x0000149f662c7301 in poll () from /lib64/libc.so.6
#1  0x0000149f61c8c0bf in full_read.constprop () from /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36680/CMSSW_14_0_X_2023-12-28-1100/lib/el8_amd64_gcc12/pluginFWCoreServicesPlugins.so
#2  0x0000149f61c405dc in edm::service::InitRootHandlers::stacktraceFromThread() () from /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36680/CMSSW_14_0_X_2023-12-28-1100/lib/el8_amd64_gcc12/pluginFWCoreServicesPlugins.so
#3  0x0000149f61c40f40 in sig_dostack_then_abort () from /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36680/CMSSW_14_0_X_2023-12-28-1100/lib/el8_amd64_gcc12/pluginFWCoreServicesPlugins.so
#4  <signal handler called>
#5  0x0000149f12dead44 in DTSegmentAnalysisTest::performClientDiagnostic(dqm::implementation::IGetter&) () from /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02817/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_X_2023-12-27-1100/lib/el8_amd64_gcc12/pluginDQMDTMonitorClient.so
#6  0x0000149f12d77b11 in non-virtual thunk to DQMEDHarvester::endProcessBlockProduce(edm::ProcessBlock&) () from /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02817/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_X_2023-12-27-1100/lib/el8_amd64_gcc12/pluginDQMDTMonitorClient.so
#7  0x0000149f68d03f2b in edm::one::EDProducerBase::doEndProcessBlock(edm::ProcessBlockPrincipal const&, edm::ModuleCallingContext const*) () from /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36680/CMSSW_14_0_X_2023-12-28-1100/lib/el8_amd64_gcc12/libFWCoreFramework.so
#8  0x0000149f68cee1c0 in edm::WorkerT<edm::one::EDProducerBase>::implDoEndProcessBlock(edm::ProcessBlockPrincipal const&, edm::ModuleCallingContext const*) () from /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36680/CMSSW_14_0_X_2023-12-28-1100/lib/el8_amd64_gcc12/libFWCoreFramework.so
#9  0x0000149f68c41fa2 in std::__exception_ptr::exception_ptr edm::Worker::runModuleAfterAsyncPrefetch<edm::OccurrenceTraits<edm::ProcessBlockPrincipal, (edm::BranchActionType)3> >(std::__exception_ptr::exception_ptr, edm::OccurrenceTraits<edm::ProcessBlockPrincipal, (edm::BranchActionType)3>::TransitionInfoType const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::ProcessBlockPrincipal, (edm::BranchActionType)3>::Context const*) () from /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36680/CMSSW_14_0_X_2023-12-28-1100/lib/el8_amd64_gcc12/libFWCoreFramework.so
#10 0x0000149f68c4586e in edm::SerialTaskQueue::QueuedTask<edm::SerialTaskQueueChain::push<edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::ProcessBlockPrincipal, (edm::BranchActionType)3> >::execute()::{lambda()#1}&>(tbb::detail::d1::task_group&, edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::ProcessBlockPrincipal, (edm::BranchActionType)3> >::execute()::{lambda()#1}&)::{lambda()#1}>::execute() () from /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36680/CMSSW_14_0_X_2023-12-28-1100/lib/el8_amd64_gcc12/libFWCoreFramework.so
#11 0x0000149f68e391d0 in tbb::detail::d1::function_task<edm::SerialTaskQueue::spawn(edm::SerialTaskQueue::TaskBase&)::{lambda()#1}>::execute(tbb::detail::d1::execution_data&) () from /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02817/el8_amd64_gcc12/cms/cmssw/CMSSW_14_0_X_2023-12-27-1100/lib/el8_amd64_gcc12/libFWCoreConcurrency.so
#12 0x0000149f6743a281 in tbb::detail::r1::task_dispatcher::local_wait_for_all<false, tbb::detail::r1::external_waiter> (waiter=..., t=<optimized out>, this=0x149f64bbbe00) at /data/cmsbld/jenkins/workspace/auto-builds/CMSSW_14_0_0_pre2-el8_amd64_gcc12/build/CMSSW_14_0_0_pre2-build/BUILD/el8_amd64_gcc12/external/tbb/v2021.9.0-c32bb0d84c2df4c9babc5f80ea81b5b2/tbb-v2021.9.0/src/tbb/task_dispatcher.h:322
#13 tbb::detail::r1::task_dispatcher::local_wait_for_all<tbb::detail::r1::external_waiter> (waiter=..., t=<optimized out>, this=0x149f64bbbe00) at /data/cmsbld/jenkins/workspace/auto-builds/CMSSW_14_0_0_pre2-el8_amd64_gcc12/build/CMSSW_14_0_0_pre2-build/BUILD/el8_amd64_gcc12/external/tbb/v2021.9.0-c32bb0d84c2df4c9babc5f80ea81b5b2/tbb-v2021.9.0/src/tbb/task_dispatcher.h:458
#14 tbb::detail::r1::task_dispatcher::execute_and_wait (t=<optimized out>, wait_ctx=..., w_ctx=...) at /data/cmsbld/jenkins/workspace/auto-builds/CMSSW_14_0_0_pre2-el8_amd64_gcc12/build/CMSSW_14_0_0_pre2-build/BUILD/el8_amd64_gcc12/external/tbb/v2021.9.0-c32bb0d84c2df4c9babc5f80ea81b5b2/tbb-v2021.9.0/src/tbb/task_dispatcher.cpp:168
#15 0x0000149f68c0379b in edm::FinalWaitingTask::wait() () from /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36680/CMSSW_14_0_X_2023-12-28-1100/lib/el8_amd64_gcc12/libFWCoreFramework.so
#16 0x0000149f68c08c4b in edm::EventProcessor::endProcessBlock(bool, bool) () from /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36680/CMSSW_14_0_X_2023-12-28-1100/lib/el8_amd64_gcc12/libFWCoreFramework.so
#17 0x0000149f68c0d781 in edm::EventProcessor::runToCompletion() () from /cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43522/36680/CMSSW_14_0_X_2023-12-28-1100/lib/el8_amd64_gcc12/libFWCoreFramework.so
#18 0x00000000004074ef in tbb::detail::d1::task_arena_function<main::{lambda()#1}::operator()() const::{lambda()#1}, void>::operator()() const ()
#19 0x0000149f674269ad in tbb::detail::r1::task_arena_impl::execute (ta=..., d=...) at /data/cmsbld/jenkins/workspace/auto-builds/CMSSW_14_0_0_pre2-el8_amd64_gcc12/build/CMSSW_14_0_0_pre2-build/BUILD/el8_amd64_gcc12/external/tbb/v2021.9.0-c32bb0d84c2df4c9babc5f80ea81b5b2/tbb-v2021.9.0/src/tbb/arena.cpp:688
#20 0x0000000000408ed2 in main::{lambda()#1}::operator()() const ()
#21 0x000000000040517c in main ()

Current Modules:

Module: DTSegmentAnalysisTest:segmentTest (crashed)
```
The harvesting step is run with one thread even in the THREADED tests (the preceding steps of the workflow were run with 4 threads).

makortel commented:
> Unless I missed something, Matti didn't find any problems outside the Framework with the new behavior.

I came to the conclusion that `DQMStore` should be safe (although given the crash I'm starting to doubt my earlier conclusion).

I listed here https://github.com/makortel/framework/issues/703 modules whose behavior should be checked. IIRC one of them looked like it would need to be fixed, the rest _might_ be fine (but should still be checked).

wddgit commented:
I'll look at the seg fault and generally start looking at the impact of this on code outside the Core today and the things listed in makortel/framework#703

makortel commented:
@cmsbuild, please test

Let's see how reproducible the crash i as well

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ef83f9/36692/summary.html
**COMMIT**: 44994e6e369e4f6aa5fc61c7bd59ac478ab931e9
**CMSSW**: CMSSW_14_0_X_2024-01-02-1100/el8_amd64_gcc12
**Additional Tests**: THREADING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43522/36692/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 27 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-02-1100+ef83f9/60446/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3247277
 - DQMHistoTests: [Total failures: 1219](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-02-1100+ef83f9/60446/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3246036
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-02-1100+ef83f9/60446/triggerResults): no differences found

wddgit commented:
Didn't reproduce. So far it has failed to reproduce in my local working area also. One thing I noticed is that in the failing run of the 5 steps in 7.3 is that I see this at step 4 in the JobReport:
```
<PerformanceSummary Metric="ProcessingSummary">
<Metric Name="NumberBeginLumiCalls" Value="0"/>
<Metric Name="NumberBeginRunCalls" Value="1"/>
<Metric Name="NumberEvents" Value="0"/>
</PerformanceSummary>
```

And in the one that passed at step 4 I see:
```
<PerformanceSummary Metric="ProcessingSummary">
<Metric Name="NumberBeginLumiCalls" Value="1"/>
<Metric Name="NumberBeginRunCalls" Value="1"/>
<Metric Name="NumberEvents" Value="1"/>
</PerformanceSummary>
```
Step 5 is the one that fails, which uses DQMRootSource, input file step3_inDQM.root. I'm not sure what all that means yet. Continuing to look at this.

makortel commented on FWCore/Framework/src/LuminosityBlockProcessingStatus.h:
```diff
@@ -103,10 +105,13 @@ namespace edm {
-    std::atomic<unsigned int> nStreamsStillProcessingLumi_{0};  //read/write as streams finish lumi so must be atomic
     edm::Timestamp endTime_{};
-    std::atomic<char> endTimeSetStatus_{0};
+    unsigned int nStreamsProcessingLumi_{0};
```
The updates are protected by the spinloops on `state_`, right? How about
```suggestion
    CMS_THREAD_GUARD(state_) unsigned int nStreamsProcessingLumi_{0};
```
?

wddgit replied:
I did this. I assume this is for the static analyzer. I don't understand when this is needed very well. It doesn't hurt anything so I just added it. The variable is only written or read in two non-const functions. Is there a specific reason the static analyzer will complain if this is not added here?

makortel replied:
In this case, where the variable is modified in non-`const` functions, and therefore is not declared as `mutable`, the static analyzer indeed would not flag it. I'm more thinking of documenting it for human reader (in which case the information could be conveyed also in comments). E.g. it took me a while to realize the updates are protected by the `state_` rather than the `SerialTaskQueue`s.

makortel commented on FWCore/Framework/src/EventProcessor.cc:
```diff
@@ -2256,6 +2260,7 @@ namespace edm {
                 // If appropriate, this will also start the next run.
                 endRunAsync(streamRunStatus_[iStreamIndex], iTask);
               }
+              status->noMoreEventsInLumi();
```
Just thinking out loud for a possible (logical) data race.

The `noMoreEventsInLumi()` sets the `LuminosityBlockProcessingStatus::nStreamsStillProcessingLumi_`, is called after the `startNextLumiOrEndRun()`.

**If** some other thread would call the `handleNextEventForStreamAsync()` for the same stream before this thread calls `noMoreEventsInLumi()`, but after it called `startNextLumiOrEndRun()`, the other thread would call `streamEndLumiAsync()`, where it would see `status->streamFinishedLumi()` returning `true` (because the status->nStreamsStillProcessingLumi_` would still be zero at that point).

It is not clear to me though if the situation above could happen. The `handleNextEventForStreamAsync()` is called from `beginLumiAsync()`, from a task queued in `streamQueues_[iStreamIndex]`, and only after the task paused the `streamQueues_[iStreamIndex]`.

I might well be missing something in either direction.

wddgit replied:
This line needs to be moved just above line 2247. That was the bug (new in this PR) causing the seg fault. If line 2220 executes before this on another stream, the atomic counter is decremented before being set and global begin lumi gets skipped. I'll push a fix with this change soon along with Matti's suggested changes.

wddgit commented:
I just noticed your second comment. Independently, I noticed the problem with the position of the call to ```noMoreEventsInLumi```.  I guess the messages crossed just before I went to lunch. Anyway, the first paragraph correctly identifies the problem. The second and third paragraphs aren't quite describing the consequences correctly. I don't think it could be called again for the same stream, but it is in fact a problem if the atomic counter is decremented before it is set.

I'll run the test a few times before I push this. It could only reproduce less than half the time in test 7.3 (and most tests didn't hit the problem).

wddgit commented on FWCore/Framework/src/LuminosityBlockProcessingStatus.cc:
```diff
@@ -26,6 +26,36 @@ namespace edm {
+    if (state_ == State::kNoMoreEvents)
+      return false;
+
+    auto expected = State::kRunning;
```
Done, that is better. I push these after I repeat the tests a few more times.

wddgit commented on FWCore/Framework/src/LuminosityBlockProcessingStatus.cc:
```diff
@@ -26,6 +26,36 @@ namespace edm {
+
+  void LuminosityBlockProcessingStatus::noMoreEventsInLumi() {
+    auto expected = State::kRunning;
+    bool changed = false;
```
Done.

wddgit commented on FWCore/Framework/src/LuminosityBlockProcessingStatus.cc:
```diff
@@ -26,6 +26,36 @@ namespace edm {
+    auto expected = State::kRunning;
+    bool changed = false;
+    do {
+      changed = state_.compare_exchange_strong(expected, State::kUpdating);
+      expected = State::kRunning;
```
Done.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43522/38311

- This PR adds an extra 92KB to repository

cmsbuild commented:
Pull request #43522 was updated. @smuzaffar, @cmsbuild, @makortel, @Dr15Jones can you please check and sign again.

wddgit commented:
please test

Rebased, fixed the bug, plus made suggested changes in Matti's second round of comments. Test 7.3 passed 6 times locally. Hopefully, all fixed now. I'll run it a few more times locally in parallel with the PR tests.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ef83f9/36704/summary.html
**COMMIT**: 03a968c74dc8d4fb170369a608afa4ac3d5c4922
**CMSSW**: CMSSW_14_0_X_2024-01-03-1100/el8_amd64_gcc12
**Additional Tests**: THREADING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43522/36704/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 85 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-03-1100+ef83f9/60458/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 22 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-03-1100+ef83f9/60458/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3247277
 - DQMHistoTests: [Total failures: 1219](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-03-1100+ef83f9/60458/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3246036
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-03-1100+ef83f9/60458/triggerResults): no differences found

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_14_1_X. Please open a backport if it should also go in to CMSSW_14_0_X.

makortel commented:
All the PRs to improve client code that was identified to be affected by this change have been merged, so we can proceed towards the integration of this PR. I'll wait for the next IB before rerunning the tests.

@wddgit Could you change the status of this PR from Draft to normal?

wddgit commented:
@makortel Yes. I'll work on this today and remove the "Draft" and anything else needed to get it moving forward.

wddgit commented:
enable threading

wddgit commented:
I read through this one more time. I think it is ready as soon as we get a new IB and rerun the tests. It is not a draft anymore.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ef83f9/37855/summary.html
**COMMIT**: 03a968c74dc8d4fb170369a608afa4ac3d5c4922
**CMSSW**: CMSSW_14_1_X_2024-03-01-1100/el8_amd64_gcc12
**Additional Tests**: THREADING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43522/37855/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 11 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-03-01-1100+ef83f9/61477/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 64 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-03-01-1100+ef83f9/61477/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3338784
 - DQMHistoTests: [Total failures: 1358](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-03-01-1100+ef83f9/61477/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3337404
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 205 log files, 166 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-03-01-1100+ef83f9/61477/triggerResults): no differences found

makortel commented:
On the last attempt the comparison differences in 9.0 are gone.

makortel commented:
According to the release scheduled 14_1_0_pre1 should have been built this week. At this point I'd leave this PR to be merged soon after the pre1 has been built, just to play safe.

makortel commented:
+core

makortel commented:
hold

cmsbuild commented:
Pull request has been put on hold by  @makortel
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

wddgit commented:
I was cleaning up old branches in my GitHub account and deleted this one by mistake. Luckily I noticed and restored it immediately. Reopening PR.

makortel commented:
unhold

14_1_0_pre1 was released

makortel commented:
@cmsbuild, please test

Lets' refresh

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @antoniovilela, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ef83f9/38062/summary.html
**COMMIT**: 03a968c74dc8d4fb170369a608afa4ac3d5c4922
**CMSSW**: CMSSW_14_1_X_2024-03-12-1100/el8_amd64_gcc12
**Additional Tests**: THREADING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43522/38062/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 101 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-03-12-1100+ef83f9/61621/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 56 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-03-12-1100+ef83f9/61621/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3297383
 - DQMHistoTests: [Total failures: 1352](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-03-12-1100+ef83f9/61621/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3296011
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 202 log files, 165 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-03-12-1100+ef83f9/61621/triggerResults): no differences found

wddgit commented:
I just force pushed. The code should be exactly the same. I squashed and rebased. The history will be a little cleaner.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43522/39460

- This PR adds an extra 12KB to repository

cmsbuild commented:
Pull request #43522 was updated. @cmsbuild, @makortel, @Dr15Jones, @smuzaffar can you please check and sign again.

wddgit commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ef83f9/38109/summary.html
**COMMIT**: f89e62114e787a0f0f7f689347bfd9aee1e4349b
**CMSSW**: CMSSW_14_1_X_2024-03-13-1100/el8_amd64_gcc12
**Additional Tests**: THREADING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43522/38109/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 103 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-03-13-1100+ef83f9/61651/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 53 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-03-13-1100+ef83f9/61651/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3297383
 - DQMHistoTests: [Total failures: 9](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-03-13-1100+ef83f9/61651/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3297354
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 202 log files, 165 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-03-13-1100+ef83f9/61651/triggerResults): no differences found

makortel commented:
+core

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @antoniovilela, @rappoccio, @sextonkennedy (and backports should be raised in the release meeting by the corresponding L2)

makortel commented:
I'd suggest to not merge this PR on Friday because of slight risk of breaking something. Merging today or Sunday/Monday would be preferred.

rappoccio commented:
+1
